### PR TITLE
feat: update entity provider schedulers

### LINKED ISSUE #816: Plugin Entity Provider schedules should be configurable through the app configurations even with schedule already defined in code
### What do you want to improve?
The plugins using an Entity Provider should have the Entity Provider schedule provided from the `app-config.yaml` take precedence over the hardcoded schedule in the code.
The current plugins using an entity provider in our repository are:
- [3scale-backend](https://github.com/janus-idp/backstage-plugins/blob/94a81885b3893c287772d4b4c53aca37f1096c0b/plugins/3scale-backend/src/providers/ThreeScaleApiEntityProvider.ts#L61-L72)
- [ocm-backend](https://github.com/janus-idp/backstage-plugins/blob/94a81885b3893c287772d4b4c53aca37f1096c0b/plugins/ocm-backend/src/providers/ManagedClusterProvider.ts#L85-L87)
- [aap-backend](https://github.com/janus-idp/backstage-plugins/blob/94a81885b3893c287772d4b4c53aca37f1096c0b/plugins/aap-backend/src/providers/AapResourceEntityProvider.ts#L46-L52)
- [keycloak-backend](https://github.com/janus-idp/backstage-plugins/blob/94a81885b3893c287772d4b4c53aca37f1096c0b/plugins/keycloak-backend/src/providers/KeycloakOrgEntityProvider.ts#L135-L137)

Additionally, a method to manually trigger the entity providers outside their scheduled times should exist.

### What is the current behavior?
Currently all of the above plugins' entity provider schedules can't be configured from the `app-config.yaml`'s `catalog.providers.<plugin>.schedule` field.
### What will the new behavior be?
Schedules from the `app-config.yaml` should override the schedules hardcoded into the backstage instance when installing the entity providers.

Additionally, a method to manually trigger the entity providers outside their scheduled times should exist.

### PR DESCRIPTION
**Pull request by Zaperex**
## Description
Updates the entity providers for 3scale-backend, ocm-backend, aap-backend and keycloak-backend to prioritize the `scheduler` option over the `schedule` option when both are provided.

## What issue(s) does this fix?
Fixes #816

### CODE REVIEW

**Comment by Zaperex** [comment]
@davidfestal this shouldn't affect the dynamic plugins right? All this does is make the `app-config.yaml` schedules (if they exist) to be prioritized over the hard coded schedules, if both `schedule` and `scheduler` are provided.

**Review by invincibleJai — APPROVED** [review]
LGTM

File: plugins/aap-backend/README.md [hunk 1]
```diff
@@ -63,7 +63,15 @@ The AAP Backstage provider plugin allows the configuration of one or multiple pr
      }
      ```
 
-   - Add a schedule directly inside the `packages/backend/src/plugins/catalog.ts` file as follows:
+     ***
+
+     **NOTE**
+
+     If any changes to the schedule in the `app-config.yaml` are made, a restart of the backend is necessary to apply the changes.
```
Comment (hmanwani-rh) [inline plugins/aap-backend/README.md:70]:
```suggestion
     If you have made any changes to the schedule in the `app-config.yaml` file, then restart to apply the changes.
```

File: plugins/aap-backend/README.md [hunk 2]
```diff
@@ -63,7 +63,15 @@ The AAP Backstage provider plugin allows the configuration of one or multiple pr
...
+
+     ***
+
+   - **Method 2**: Add a schedule directly inside the `packages/backend/src/plugins/catalog.ts` file as follows:
```
Comment (hmanwani-rh) [inline plugins/aap-backend/README.md:74]:
```suggestion
   - Add a schedule directly inside the `packages/backend/src/plugins/catalog.ts` file as follows:
```

File: plugins/aap-backend/README.md [hunk 3]
```diff
@@ -93,6 +101,39 @@ The AAP Backstage provider plugin allows the configuration of one or multiple pr
      }
      ```
 
+   - If both the `schedule` (hard-coded schedule) and `scheduler` (`app-config.yaml` schedule) option are provided in the `packages/backend/src/plugins/catalog.ts`, the `scheduler` option takes precedence.
+
+     - If the schedule inside the `app-config.yaml` is not configured, then the `schedule` option is used.
```
Comment (hmanwani-rh) [inline plugins/aap-backend/README.md:106]:
```suggestion
     - If the schedule inside the `app-config.yaml` file is not configured, then the `schedule` option is used.
```

File: plugins/keycloak-backend/README.md [hunk 4]
```diff
@@ -85,10 +85,21 @@ yarn workspace backend add @janus-idp/backstage-plugin-keycloak-backend
      }
      ```
 
+     ***
+
+     **NOTE**
+
+     If any changes to the schedule in the `app-config.yaml` are made, a restart of the backend is necessary to apply the changes.
```
Comment (hmanwani-rh) [inline plugins/keycloak-backend/README.md:92]:
```suggestion
     If you have made any changes to the schedule in the `app-config.yaml` file, then restart to apply the changes.
```

File: plugins/keycloak-backend/README.md [hunk 5]
```diff
@@ -104,11 +115,48 @@ yarn workspace backend add @janus-idp/backstage-plugin-keycloak-backend
...
+
+       const { processingEngine, router } = await builder.build();
+       await processingEngine.start();
+       return router;
+     }
+     ```
+
+   - If both the `schedule` (hard-coded schedule) and `scheduler` (`app-config.yaml` schedule) option are provided in the `packages/backend/src/plugins/catalog.ts`, the `scheduler` option takes precedence.
+
+     - If the schedule inside the `app-config.yaml` is not configured, then the `schedule` option is used.
```
Comment (hmanwani-rh) [inline plugins/keycloak-backend/README.md:132]:
```suggestion
     - If the schedule inside the `app-config.yaml` file is not configured, then the `schedule` option is used.
```

File: plugins/ocm/README.md [hunk 6]
```diff
@@ -166,69 +166,105 @@ If you are interested in Resource discovery and do not want any of the front-end
...
-            frequency: { minutes: 1 },
-            timeout: { minutes: 1 },
-          }),
-        });
-        builder.addEntityProvider(ocm);
-        /* highlight-add-start */
-        // ...
-      }
-      ```
+   - **Method 1**: Configure the scheduler inside the `app-config.yaml`:
```
Comment (hmanwani-rh) [inline plugins/ocm/README.md:169]:
```suggestion
   - * Configure the scheduler inside the `app-config.yaml` file:
```

File: plugins/ocm/README.md [hunk 7]
```diff
@@ -166,69 +166,105 @@ If you are interested in Resource discovery and do not want any of the front-end
...
+       /* highlight-add-start */
+       // ...
+     }
+     ```
+
+     ***
+
+     **NOTE**
+
+     If any changes to the schedule in the `app-config.yaml` are made, a restart of the backend is necessary to apply the changes.
```
Comment (hmanwani-rh) [inline plugins/ocm/README.md:212]:
```suggestion
     If you have made any changes to the schedule in the `app-config.yaml` file, then restart to apply the changes.
```

File: plugins/ocm/README.md [hunk 8]
```diff
@@ -166,69 +166,105 @@ If you are interested in Resource discovery and do not want any of the front-end
...
+
+     ***
+
+   - **Method 2**: Add a schedule directly inside the `packages/backend/src/plugins/catalog.ts` file:
```
Comment (hmanwani-rh) [inline plugins/ocm/README.md:216]:
```suggestion
   - Add a schedule directly inside the `packages/backend/src/plugins/catalog.ts` file:
```

File: plugins/ocm/README.md [hunk 9]
```diff
@@ -166,69 +166,105 @@ If you are interested in Resource discovery and do not want any of the front-end
...
+       });
+       builder.addEntityProvider(ocm);
+       /* highlight-add-start */
+       // ...
+     }
+     ```
+
+   - If both the `schedule` (hard-coded schedule) and `scheduler` (`app-config.yaml` schedule) option are provided in the `packages/backend/src/plugins/catalog.ts`, the `scheduler` option takes precedence.
+
+     - If the schedule inside the `app-config.yaml` is not configured, then the `schedule` option is used.
```
Comment (hmanwani-rh) [inline plugins/ocm/README.md:243]:
```suggestion
     - If the schedule inside the `app-config.yaml` file is not configured, then the `schedule` option is used.
```

File: plugins/3scale-backend/README.md [hunk 10]
```diff
@@ -34,32 +34,105 @@ yarn workspace backend add @janus-idp/backstage-plugin-3scale-backend
...
+
+       const { processingEngine, router } = await builder.build();
+       await processingEngine.start();
+       return router;
+     }
+     ```
+
+   - If both the `schedule` (hard-coded schedule) and `scheduler` (`app-config.yaml` schedule) option are provided in the `packages/backend/src/plugins/catalog.ts`, the `scheduler` option takes precedence.
+
+     - If the schedule inside the `app-config.yaml` is not configured, then the `schedule` option is used.
```
Comment (hmanwani-rh) [inline plugins/3scale-backend/README.md:106]:
```suggestion
     - If the schedule inside the `app-config.yaml` file is not configured, then the `schedule` option is used.
```

**Review by hmanwani-rh — CHANGES_REQUESTED** [review]
@Zaperex I've added a few comments below. Please take a look and let me know if you've any questions.

File: plugins/3scale-backend/README.md [hunk 11]
```diff
@@ -34,32 +34,105 @@ yarn workspace backend add @janus-idp/backstage-plugin-3scale-backend
...
-     /* highlight-add-end */
-
-     const { processingEngine, router } = await builder.build();
-     await processingEngine.start();
-     return router;
-   }
-   ```
+2. Configure the scheduler for the entity provider using one of the following methods:
+
+   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` file, add the following code to `packages/backend/src/plugins/catalog.ts` file:
```
[the remark this answers is not in the document]
Comment (Zaperex) [reply plugins/3scale-backend/README.md:39]: @hmanwani-rh so the reasoning to explicitly mention `Method 1` and `Method 2` was to make it obvious that there are 2 explicit methods of configuring the scheduler (and that they don't need to use both), and then mention the behavior when both methods are used simultaneously in a section below them.

File: plugins/3scale-backend/README.md [hunk 12]
```diff
@@ -34,32 +34,105 @@ yarn workspace backend add @janus-idp/backstage-plugin-3scale-backend
...
+
+     ***
+
+     **NOTE**
+
+     If any changes to the schedule in the `app-config.yaml` are made, a restart of the backend is necessary to apply the changes.
+
+     ***
+
+   - **Method 2**: Add a schedule directly inside the `packages/backend/src/plugins/catalog.ts` file as follows:
```
[the remark this answers is not in the document]
Comment (Zaperex) [reply plugins/3scale-backend/README.md:74]: Same as [above](https://github.com/janus-idp/backstage-plugins/pull/827#discussion_r1352722633)

File: plugins/aap-backend/README.md [hunk 13]
```diff
@@ -36,7 +36,7 @@ The AAP Backstage provider plugin allows the configuration of one or multiple pr
 
 1. Configure the scheduler using one of the following options:
 
-   - Add the following code to the `packages/backend/src/plugins/catalog.ts` file if the scheduler is configured inside the `app-config.yaml` file:
+   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` file, you can use it by adding the following code to the `packages/backend/src/plugins/catalog.ts` file:
```
[the remark this answers is not in the document]
Comment (Zaperex) [reply plugins/aap-backend/README.md:39]: Same as [above](https://github.com/janus-idp/backstage-plugins/pull/827#discussion_r1352722633)

File: plugins/aap-backend/README.md (hunk 2 above)
Comment (Zaperex) [reply plugins/aap-backend/README.md:74]: Same as [above](https://github.com/janus-idp/backstage-plugins/pull/827#discussion_r1352722633)

File: plugins/ocm/README.md (hunk 8 above)
Comment (Zaperex) [reply plugins/ocm/README.md:216]: Same as [above](https://github.com/janus-idp/backstage-plugins/pull/827#discussion_r1352722633)

File: plugins/ocm/README.md (hunk 6 above)
Comment (Zaperex) [reply plugins/ocm/README.md:169]: Same as [above](https://github.com/janus-idp/backstage-plugins/pull/827#discussion_r1352722633)

File: plugins/3scale-backend/README.md [hunk 14]
```diff
@@ -34,32 +34,105 @@ yarn workspace backend add @janus-idp/backstage-plugin-3scale-backend
...
+
+       const { processingEngine, router } = await builder.build();
+       await processingEngine.start();
+       return router;
+     }
+     ```
+
+   - If both the `schedule` (hard-coded schedule) and `scheduler` (`app-config.yaml` schedule) option are provided in the `packages/backend/src/plugins/catalog.ts`, the `scheduler` option takes precedence.
+
+     - If the schedule inside the `app-config.yaml` file is not configured, then the `schedule` option is used.
```
Comment (tumido) [inline plugins/3scale-backend/README.md:106]: I do not understand why is this here. You're within a section that says "both are defined", so why backtrack to the "Method 2" in this sentence?

File: plugins/3scale-backend/README.md [hunk 15]
```diff
@@ -34,32 +34,105 @@ yarn workspace backend add @janus-idp/backstage-plugin-3scale-backend
...
+           scheduler: env.scheduler,
+         }),
+       );
+       /* highlight-add-end */
+
+       const { processingEngine, router } = await builder.build();
+       await processingEngine.start();
+       return router;
+     }
+     ```
```
Comment (tumido) [inline plugins/3scale-backend/README.md:135]: Isn't this just a repetition?

Besides, having both `scheduler` and `schedule` defined is an antipattern that should not be encouraged.

**Review by tumido — COMMENTED** [review]
Code LGTM. The documentation feels a bit more confusing to me than it was before.

File: plugins/3scale-backend/README.md (hunk 14 above)
Comment (Zaperex) [reply plugins/3scale-backend/README.md:106]: I mainly wanted to mention the expected behavior when both are `schedule` and `scheduler` are defined.

File: plugins/3scale-backend/README.md (hunk 15 above)
Comment (Zaperex) [reply plugins/3scale-backend/README.md:135]: Ah I did not know this was an antipattern. I assumed the `schedule` would serve as a hard coded default while the `scheduler` served as something the user can customize to override the hard coded default.

Comment (Zaperex) [reply plugins/3scale-backend/README.md:135]: I can remove the repeated code for when both `scheduler` and `schedule` are defined.

Comment (tumido) [reply plugins/3scale-backend/README.md:135]: Oh.. I didn't think of that... Maybe we're making it a bigger issue than it is. I would say the user should configure it in one or the other place and not both to keep it simple. The reason why there are 2 ways is that in typescript user has more options on how to do context-aware configuration if they need so (like conditionals, feature flag checks etc...) In most cases the `schedule` in `app-config.yaml` should be sufficient and preferred way.

File: plugins/3scale-backend/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (tumido) [reply plugins/3scale-backend/README.md:39]: Maybe we should clarify the wording here:


```
If the scheduler is configured inside the `app-config.yaml`...
```

The word *scheduler* here actually means `schedule` config key within `app-config.yaml`.

File: plugins/3scale-backend/README.md (hunk 15 above)
Comment (Zaperex) [reply plugins/3scale-backend/README.md:135]: Ok, so should I just leave the behavior when both are defined as a mention, but just mention it's not recommended?

File: plugins/3scale-backend/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (hmanwani-rh) [reply plugins/3scale-backend/README.md]: @Zaperex Listing all these points as bullet points might imply that the subsequent bullets are also methods regardless of specifically marking methods, as non-method bullets are part of the same list. Therefore, I recommend separating these non-method bullets into a new step. What do you think?

File: plugins/3scale-backend/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (hmanwani-rh) [reply plugins/3scale-backend/README.md:74]: Please see my previous comment.

File: plugins/aap-backend/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (hmanwani-rh) [reply plugins/aap-backend/README.md]: Please see my previous comment.

File: plugins/ocm/README.md (hunk 8 above)
Comment (hmanwani-rh) [reply plugins/ocm/README.md:216]: Please see my previous comment.

File: plugins/3scale-backend/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Zaperex) [reply plugins/3scale-backend/README.md]: @hmanwani-rh I think that could potentially help make the methods stand out more. Alternatively, we can wrap this edge case in a note block:
```markdown
***

**NOTE**

If both the `schedule` (hard-coded schedule) and `scheduler` (`app-config.yaml` schedule) option are provided in the `packages/backend/src/plugins/catalog.ts`, the `scheduler` option takes precedence. However, if the schedule inside the `app-config.yaml` file is not configured, then the `schedule` option is used.

***
```
What do you think?

[the remark this answers is not in the document]
Comment (hmanwani-rh) [reply plugins/3scale-backend/README.md]: Adding NOTE makes sense.

File: plugins/3scale-backend/README.md [hunk 16]
```diff
@@ -34,32 +34,80 @@ yarn workspace backend add @janus-idp/backstage-plugin-3scale-backend
...
-     /* highlight-add-end */
-
-     const { processingEngine, router } = await builder.build();
-     await processingEngine.start();
-     return router;
-   }
-   ```
+2. Configure the scheduler for the entity provider using one of the following methods:
+
+   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` via the schedule config key shown above, add the following code to `packages/backend/src/plugins/catalog.ts` file:
```
Comment (hmanwani-rh) [inline plugins/3scale-backend/README.md:39]:
```suggestion
   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` using the schedule config key mentioned previously, add the following code to `packages/backend/src/plugins/catalog.ts` file:
```

File: plugins/aap-backend/README.md [hunk 17]
```diff
@@ -36,7 +36,7 @@ The AAP Backstage provider plugin allows the configuration of one or multiple pr
 
 1. Configure the scheduler using one of the following options:
 
-   - Add the following code to the `packages/backend/src/plugins/catalog.ts` file if the scheduler is configured inside the `app-config.yaml` file:
+   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` via the schedule config key shown above, add the following code to `packages/backend/src/plugins/catalog.ts` file:
```
Comment (hmanwani-rh) [inline plugins/aap-backend/README.md:39]:
```suggestion
   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` using the schedule config key mentioned previously, add the following code to `packages/backend/src/plugins/catalog.ts` file:
```

File: plugins/ocm/README.md [hunk 18]
```diff
@@ -166,69 +166,85 @@ If you are interested in Resource discovery and do not want any of the front-end
...
-            frequency: { minutes: 1 },
-            timeout: { minutes: 1 },
-          }),
-        });
-        builder.addEntityProvider(ocm);
-        /* highlight-add-start */
-        // ...
-      }
-      ```
+   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` via the schedule config key shown below:
```
Comment (hmanwani-rh) [inline plugins/ocm/README.md:169]:
```suggestion
   - **Method 1**: If the scheduler is configured inside the `app-config.yaml` using the schedule config key as follows:
```

**Review by hmanwani-rh — COMMENTED** [review]
@Zaperex Added a few suggestions below. Please take a look and let me know if you've any questions.

**Review by tumido — APPROVED** [review]
/lgtm
/approve
